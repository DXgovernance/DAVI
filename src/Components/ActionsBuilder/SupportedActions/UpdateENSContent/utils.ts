import contenthash from 'content-hash';
import { MAINNET_ID, LOCALHOST_ID, GOERLI_ID } from 'utils';
import { utils } from 'ethers';

const DEFAULT_NAMEHASH =
  '0x0000000000000000000000000000000000000000000000000000000000000000';

export const convertToNameHash = (name: string) => {
  if (!name) {
    return DEFAULT_NAMEHASH;
  }
  return utils.namehash(name);
};

export const convertToContentHash = (ipfsHash: string) => {
  if (!ipfsHash) {
    return '';
  }
  return contenthash.fromIpfs(ipfsHash);
};

export const convertToIpfsHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const getIpfsUrl = (ipfsHash: string) => {
  const ipfsRoot = 'ipfs://';
  return `${ipfsRoot}${ipfsHash}`;
};

export const isValidChainId = (chainId: number) => {
  if (!chainId || chainId === LOCALHOST_ID) {
    return MAINNET_ID;
  } else {
    return chainId;
  }
};

export const isAvailableOnENS = (chainId: number) => {
  const validChainId = isValidChainId(chainId);
  const ensNetworks = [MAINNET_ID, GOERLI_ID];
  return ensNetworks.includes(validChainId);
};
