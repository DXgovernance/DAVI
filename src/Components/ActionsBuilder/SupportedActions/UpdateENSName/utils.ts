import namehash from 'eth-ens-namehash';
import contenthash from 'content-hash';
import { useState, useEffect } from 'react';

export const convertToNameHash = (name: string) => namehash.hash(name);

export const convertToContentHash = (ipfsHash: string) =>
  contenthash.fromIpfs(ipfsHash);

export const convertToIPFSHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const useDebounce = (value: string, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
};
