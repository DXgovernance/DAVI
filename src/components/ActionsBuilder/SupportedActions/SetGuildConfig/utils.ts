import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
