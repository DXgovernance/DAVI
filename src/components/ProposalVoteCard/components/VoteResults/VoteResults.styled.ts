import styled from 'styled-components';

export const VotesRowWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  font-weight: ${({ theme }) => theme.fontWeights.regular};
  font-size: ${({ theme }) => theme.fontSizes.body};
  margin: 5px 0px 5px 0px;
`;

export const VoteOption = styled.div`
  display: flex;
  align-items: center;
`;

export const OptionBullet = styled.span`
  margin-right: 0.5rem;
`;
