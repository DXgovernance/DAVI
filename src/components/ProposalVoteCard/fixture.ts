import { BigNumber } from 'ethers';
import moment from 'moment';
import { ContractState } from 'types/types.guilds.d';

const mockBignumber = BigNumber.from(100000000);
const mockBigNumberOption = BigNumber.from(100000000000000);
const mockBigNumberTotalSupply = BigNumber.from(1000000000000000);
const mockBigNumberTotalLocked = BigNumber.from(1000000000000000);

export const mockProposalVoteCardProps = {
  proposal: {
    id: '0x0',
    metadata: {
      description: 'This is a test proposal',
      voteOptions: ['Yes', 'No'],
    },
    endTime: moment(),
    creator: '0x1234567890123456789012345678901234567890',
    startTime: moment(),
    timeDetail: '',
    title: 'SWPR single reward campaign',
    to: [],
    value: [],
    totalOptions: BigNumber.from(0),
    contentHash: '0x1234567890123456789012345678901234567890',
    contractState: ContractState.Active,
    totalVotes: [],
  },
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
  timestamp: 10,
  votingPower: {
    percent: 100,
    userVotingPower: mockBignumber,
    atSnapshot: mockBignumber,
    atCurrentSnapshot: mockBignumber,
  },
  contract: {} as any,
  createTransaction: () => {},
  currentLockedPercent: 100,
  userVote: {
    option: '0',
    votingPower: BigNumber.from(100000000),
  },
};

export const mockVoteResults = {
  isPercent: true,
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
  proposalMetadata: {
    description: 'This is a test proposal',
    voteOptions: ['Yes', 'No'],
  },
};

export const mockVoteConfirmationModal = {
  isOpen: true,
  onDismiss: () => {},
  onConfirm: () => {},
  selectedAction: 'Yes',
  votingPower: 100,
  totalLocked: 100,
  onAddToVoteCart: () => {},
};

export const mockVoteChart = {
  isPercent: true,
  voteData: {
    options: {
      0: mockBigNumberOption,
      1: mockBigNumberOption,
    },
    quorum: mockBignumber,
    totalLocked: mockBigNumberTotalLocked,
    token: {
      name: 'DXdao',
      symbol: 'DXD',
      decimals: 18,
      totalSupply: mockBigNumberTotalSupply,
    },
  },
};
