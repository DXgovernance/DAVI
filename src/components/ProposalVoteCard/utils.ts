import { ConfirmVoteProposalProps, VoteOnProposalProps } from './types';

export const voteOnProposal = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};

export const confirmVoteProposal = ({
  proposal,
  contract,
  selectedOption,
  userVotingPower,
  createTransaction,
  cb,
}: ConfirmVoteProposalProps) => {
  createTransaction(
    `Vote on proposal ${proposal?.title}`,
    async () => contract.setVote(proposal?.id, selectedOption, userVotingPower),
    true,
    cb
  );
};

export const getOptionLabel = ({ metadata, optionKey, t }) => {
  const metadataLabel = metadata?.voteOptions?.[optionKey];
  return metadataLabel
    ? metadataLabel
    : Number(optionKey) === 0
    ? t('against', { defaultValue: 'Against' })
    : t('option', { optionKey });
};
