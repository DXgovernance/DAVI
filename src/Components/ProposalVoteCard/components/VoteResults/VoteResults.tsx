import { formatUnits } from 'ethers/lib/utils';
import useVotingPowerPercent from 'hooks/Guilds/guild/useVotingPowerPercent';
import Bullet from 'old-components/Guilds/common/Bullet';
import { Loading } from 'Components/Primitives/Loading';
import { useTheme } from 'styled-components';
import {
  ResultRowProps,
  VoteResultsProps,
} from 'Components/ProposalVoteCard/types';
import {
  VotesRowWrapper,
  VoteOption,
  OptionBullet,
} from './VoteResults.styled';

const VoteResultRow: React.FC<ResultRowProps> = ({
  isPercent,
  optionKey,
  voteData,
  proposalMetadata,
}) => {
  const theme = useTheme();

  const isReady = optionKey !== undefined;

  const votingPowerPercent = useVotingPowerPercent(
    voteData?.options?.[optionKey],
    voteData?.totalLocked,
    2
  );

  return (
    <VotesRowWrapper>
      <VoteOption>
        <OptionBullet>
          {isReady ? (
            <Bullet color={theme?.colors?.votes?.[optionKey]} size={8} />
          ) : (
            <Loading
              loading
              text
              skeletonProps={{ circle: true, height: 16, width: 16 }}
            />
          )}
        </OptionBullet>

        {isReady ? (
          proposalMetadata?.voteOptions?.[optionKey] ||
          'Option ' + (optionKey + 1)
        ) : (
          <Loading loading text />
        )}
      </VoteOption>
      {isReady && voteData ? (
        <span>
          {isPercent
            ? `${formatUnits(voteData?.options?.[optionKey] || 0)} ${
                voteData?.token?.symbol
              }`
            : `${votingPowerPercent}%`}
        </span>
      ) : (
        <Loading loading text skeletonProps={{ width: 50 }} />
      )}
    </VotesRowWrapper>
  );
};

const VoteResults: React.FC<VoteResultsProps> = ({
  isPercent,
  voteData,
  proposalMetadata,
}) => {
  return voteData ? (
    <>
      {Object.entries(voteData.options).map((_, i) => (
        <VoteResultRow
          optionKey={i}
          isPercent={isPercent}
          voteData={voteData}
          proposalMetadata={proposalMetadata}
        />
      ))}
    </>
  ) : (
    <>
      <VoteResultRow
        isPercent={isPercent}
        voteData={voteData}
        proposalMetadata={proposalMetadata}
      />
      <VoteResultRow
        isPercent={isPercent}
        voteData={voteData}
        proposalMetadata={proposalMetadata}
      />
    </>
  );
};

export default VoteResults;
