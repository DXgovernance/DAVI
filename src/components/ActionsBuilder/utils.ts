import { DecodedAction } from './types';
import { Permission } from './types';
import Web3 from 'web3';
import { NULL_SIGNATURE } from 'utils';
import { SupportedAction } from './types';

const web3 = new Web3();

export const getPermissionArgs = (
  decodedActions: DecodedAction[]
): Permission[] => {
  const permissions = decodedActions.map(decodedAction => {
    // Native transfer of ETH does not have a function
    if (
      decodedAction?.decodedCall?.callType ===
        SupportedAction.NATIVE_TRANSFER ||
      decodedAction?.decodedCall?.callType === SupportedAction.RAW_TRANSACTION
    ) {
      return {
        from: decodedAction?.decodedCall?.from,
        to: decodedAction?.decodedCall?.to,
        callType: decodedAction?.decodedCall?.callType,
        functionSignature: NULL_SIGNATURE,
      };
    }
    const functionInputs = decodedAction?.decodedCall?.function?.inputs?.map(
      input => input.type
    );
    const functionName = decodedAction?.decodedCall?.function?.name;
    const functionSignature = web3.eth.abi.encodeFunctionSignature(
      `${functionName}(${functionInputs.join(',')})`
    );
    return {
      from: decodedAction?.decodedCall?.from,
      to: decodedAction?.decodedCall?.to,
      callType: decodedAction?.decodedCall?.callType,
      functionSignature,
    };
  });
  return permissions;
};
