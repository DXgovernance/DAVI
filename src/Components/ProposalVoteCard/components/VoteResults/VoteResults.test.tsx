import { render } from 'utils/tests';
import VoteResults from './VoteResults';
import { mockVoteResults } from '../../fixture';

jest.mock('contexts/index', () => jest.fn());

describe('VoteResults', () => {
  it('matches the snapshot', () => {
    const { container } = render(<VoteResults {...mockVoteResults} />);
    expect(container).toMatchSnapshot();
  });
});
