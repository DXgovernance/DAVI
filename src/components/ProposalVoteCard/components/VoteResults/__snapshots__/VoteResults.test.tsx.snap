// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteResults matches the snapshot 1`] = `
.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #295FF4;
  border-radius: 50%;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 8px;
  width: 8px;
  background-color: #E75C5C;
  border-radius: 50%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-weight: 500;
  font-size: 14px;
  margin: 5px 0px 5px 0px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  margin-right: 0.5rem;
}

<div>
   
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <span
          class="c3"
          color="#295FF4"
          size="8"
        />
      </span>
      No
    </div>
    <span>
      10%
    </span>
  </div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <span
          class="c4"
          color="#E75C5C"
          size="8"
        />
      </span>
      Yes
    </div>
    <span>
      10%
    </span>
  </div>
   
</div>
`;
